File: locust/cart_management.py
from locust import HttpUser, between, task
from common import EMAIL, PASSWORD, ProductionTrafficShape, auth_headers, find_product


class CartManagement(HttpUser):
    wait_time = between(1, 3)

    @task
    def cart(self):
        login = self.client.post("/api/auth/login", json={"email": EMAIL, "password": PASSWORD}, headers=auth_headers(), name="POST /api/auth/login")
        data = login.json() if login.status_code == 200 else {}
        token = data.get("tokens", {}).get("accessToken")
        refresh = data.get("tokens", {}).get("refreshToken")
        product_id = find_product(self.client)
        if product_id:
            self.client.get(f"/api/products/{product_id}", headers=auth_headers(), name="GET /api/products/:id")
            self.client.get("/api/cart", headers=auth_headers(token), name="GET /api/cart")
            response = self.client.post("/api/cart/items", json={"productId": product_id, "quantity": 1}, headers=auth_headers(token), name="POST /api/cart/items")
            item_id = response.json().get("data", {}).get("id") if response.status_code == 201 else None
            if item_id: self.client.delete(f"/api/cart/items/{item_id}", headers=auth_headers(token), name="DELETE /api/cart/items/:id")
        if refresh: self.client.post("/api/auth/logout", json={"email": EMAIL, "password": PASSWORD, "refreshToken": refresh}, headers=auth_headers(), name="POST /api/auth/logout")


class CartShape(ProductionTrafficShape):
    peak_rate = 10
    production_users = 208

File: k6/cart-management.js
import http from 'k6/http';
import { check } from 'k6';
import { BASE_URL, EMAIL, PASSWORD, jsonParams, parseJson, scenarioOptions, think } from './common.js';

export const options = scenarioOptions(10, 208, 416);

export default function () {
  let response = http.post(`${BASE_URL}/api/auth/login`, JSON.stringify({email: EMAIL, password: PASSWORD}), jsonParams());
  const auth = parseJson(response); const token = auth.tokens?.accessToken; const refreshToken = auth.tokens?.refreshToken;
  check(response, {'login 200': (r) => r.status === 200 && !!token});
  think(1, 3);
  response = http.get(`${BASE_URL}/api/products`, jsonParams());
  const id = __ENV.PRODUCT_ID || (parseJson(response).data || [])[0]?.id;
  check(response, {'products 200': (r) => r.status === 200});
  think(1, 3);
  if (id) {
    check(http.get(`${BASE_URL}/api/products/${id}`, jsonParams()), {'detail 200': (r) => r.status === 200});
    think(1, 3);
    check(http.get(`${BASE_URL}/api/cart`, jsonParams(token)), {'cart 200': (r) => r.status === 200});
    think(1, 3);
    response = http.post(`${BASE_URL}/api/cart/items`, JSON.stringify({productId: id, quantity: 1}), jsonParams(token));
    const itemId = parseJson(response).data?.id;
    check(response, {'add item 201': (r) => r.status === 201});
    think(1, 3);
    if (itemId) check(http.del(`${BASE_URL}/api/cart/items/${itemId}`, null, jsonParams(token)), {'delete item 204': (r) => r.status === 204});
  }
  think(1, 3);
  check(http.post(`${BASE_URL}/api/auth/logout`, JSON.stringify({email: EMAIL, password: PASSWORD, refreshToken}), jsonParams()), {'logout 200': (r) => r.status === 200});
}
